File: 200. Number of Islands.cpp
class Solution {
public:
    int numIslands(vector<vector<char>>& grid) {
        if (grid.size() == 0) return 0;
        vector <vector<bool>> visited (grid.size(), vector<bool> (grid[0].size(), false));
        int num_of_island = 0;
        for (int i = 0; i < grid.size(); i++)
        {
            for (int j = 0; j < grid[0].size(); j++)
            {
                if (grid[i][j] == '1' && visited[i][j] == false)
                {
                    num_of_island ++;
                    DFS(i, j, grid, visited);
                }
            }
        }
        return num_of_island;
    }
    
    void DFS(int i, int j, vector<vector<char>>& grid, vector <vector<bool>>& visited)
    {
        visited[i][j] = true;
        if (i-1 >=0 && grid[i-1][j] == '1' && visited[i-1][j] == false)
            DFS(i-1, j, grid, visited);
        if (i+1 < grid.size() && grid[i+1][j] == '1' && visited[i+1][j] == false)
            DFS(i+1, j, grid, visited);
        if (j-1 >= 0 && grid[i][j-1] == '1' && visited[i][j-1] == false)
            DFS(i, j-1, grid, visited);
        if (j+1 < grid[0].size() && grid[i][j+1] == '1' && visited[i][j+1] == false)
            DFS(i, j+1, grid, visited);
        return;
    }
};

File: 200. Number of Islands.py
class Solution(object):
    def numIslands(self, grid):
        """
        :type grid: List[List[str]]
        :rtype: int
        """
        self.visited = [[False for j in xrange(len(grid[0]))] for i in xrange(len(grid))]
        num_of_island = 0
        for i in xrange(len(grid)):
            for j in xrange(len(grid[0])):
                if grid[i][j] == '1' and self.visited[i][j] == False:
                    num_of_island += 1
                    self.DFS(i, j, grid)
        return num_of_island
        
    def DFS(self, i, j, grid):
        self.visited[i][j] = True
        if i-1 >= 0 and grid[i-1][j] == '1' and self.visited[i-1][j] == False:
            self.DFS(i-1, j, grid)
        if i+1 < len(grid) and grid[i+1][j] == '1' and self.visited[i+1][j] == False:
            self.DFS(i+1, j, grid)
        if j-1 >= 0 and grid[i][j-1] == '1' and self.visited[i][j-1] == False:
            self.DFS(i, j-1, grid)
        if j+1 < len(grid[0]) and grid[i][j+1] == '1' and self.visited[i][j+1] == False:
            self.DFS(i, j+1, grid)
        return
